packages/perseus/src/widgets/passage/passage-markdown.jsx

```diff
@@ -1,85 +1,90 @@
 /* eslint-disable react/sort-comp */
 // @flow
+
 import SimpleMarkdown from "@khanacademy/simple-markdown";
 import * as i18n from "@khanacademy/wonder-blocks-i18n";
-import PropTypes from "prop-types";
 import * as React from "react";
 import _ from "underscore";
 
-const START_REF_PREFIX = "start-ref-";
-const END_REF_PREFIX = "end-ref-";
-const REF_STYLE = {
-    display: "inline-block",
-    width: 0,
-    visibility: "hidden",
+import type {
+    Capture,
+    Parser,
+    SingleASTNode,
+} from "@khanacademy/simple-markdown";
+
+export type ParseState = {
+    currentRef: number[],
+    useRefs: boolean,
+    lastRef: number,
+    firstSentenceRef: ?string,
+    firstQuestionRef: ?string,
+    lastFootnote: {id: number, text: string},
 };
 
-const LABEL_OUTER_STYLE = {
-    // for some reason we need these to keep the nbsp from wrapping when the
-    // inner circle/square is display: inline-block
-    display: "inline",
-    whiteSpace: "nowrap",
+type FootnoteType = {
+    id: number,
+    text: string,
 };
 
-const SQUARE_LABEL_STYLE = {
-    display: "inline-block",
-    color: "rgb(255, 255, 255)",
-    backgroundColor: "rgb(90, 90, 90)",
-    paddingLeft: 10,
-    paddingRight: 10,
-    userSelect: "none",
-    WebkitUserSelect: "none",
+type RefStartNode = {
+    ref: ?number,
+    refContent: React.Node,
 };
 
-const CIRCLE_LABEL_STYLE = {
-    display: "inline-block",
-    color: "rgb(255, 255, 255)",
-    backgroundColor: "rgb(90, 90, 90)",
-    userSelect: "none",
-    WebkitUserSelect: "none",
-    width: 22,
-    height: 22,
-    borderRadius: "50%",
-    textAlign: "center",
+type RefEndNode = {
+    ref: ?number,
 };
 
-class RefStart extends React.Component<$FlowFixMe> {
-    static propTypes = {
-        refContent: PropTypes.node.isRequired,
-    };
+type LabelNode = {
+    content: string,
+    space: boolean,
+};
+
+type HighlightNode = {
+    content: string,
+};
+
+type RefStartProps = {
+    refContent: React.Node,
+};
 
+const INITIAL_PARSE_STATE: ParseState = {
+    currentRef: [],
+    useRefs: true,
+    lastRef: 0,
+    firstSentenceRef: null,
+    firstQuestionRef: null,
+    lastFootnote: {id: 0, text: ""},
+};
+
+class RefStart extends React.Component<RefStartProps> {
     render(): React.Node {
         return <span style={REF_STYLE}>{i18n.doNotTranslate("_")}</span>;
     }
 
-    getRefContent = () => {
+    getRefContent: () => React.Node = () => {
         return this.props.refContent;
     };
 }
 
-class RefEnd extends React.Component<$FlowFixMe> {
+class RefEnd extends React.Component<{}> {
     render(): React.Node {
         return <span style={REF_STYLE}>{i18n.doNotTranslate("_")}</span>;
     }
 }
 
-const rules: $FlowFixMe = {
-    // $FlowFixMe[prop-missing]
-    // $FlowFixMe[incompatible-use]
+const rules = {
```

Comment:
I really wanted this to be `const rules: ReactRules = {` but it exploded into a cacophony of errors that I couldn't figure out.